content/zh/blog/2021/better-external-authz/index.md

```diff
@@ -0,0 +1,556 @@
+---
+title: 更好的外部授权方式集成
+subtitle: 使用 AuthorizationPolicy 将外部授权系统（例如 OPA、oauth2-proxy 等）与 Istio 进行集成
+description: AuthorizationPolicy 现在支持以 CUSTOM 自定义方式委托外部系统进行授权操作。
+publishdate: 2021-02-09
+attribution: Yangmin Zhu (Google)
+keywords: [authorization,access control,opa,oauth2]
+---
+
+## 背景 {#background}
+
+Istio 的授权策略为网格中的服务提供访问控制。它速度快、功能强大且使用广泛。
+自 Istio 1.4 首次发布以来，我们不断改进策略以使其更加灵活，
+包括 [`DENY` 操作](/zh/docs/tasks/security/authorization/authz-deny/)、
+[排除语义](/zh/docs/tasks/security/authorization/authz-deny/), 
+[`X-Forwarded-For` 头信息支持](/zh/docs/tasks/security/authorization/authz-ingress/)，
+[嵌套 JWT 声明支持](/zh/docs/tasks/security/authorization/authz-jwt/)等等。
+这些特性提高了授权策略的灵活性，但仍有许多场景无法通过该模型支持，例如：
+
+- 您拥有自己的内部授权系统，该系统无法轻松迁移或替换至授权策略中。
+
+- 您想与使用 Istio 中[底层 Envoy 配置 API](/zh/docs/reference/config/networking/envoy-filter/)
+  （例如 [Open Policy Agent](https://www.openpolicyagent.org/docs/latest/envoy-introduction/)
+  或 [`oauth2` 代理](https://github.com/oauth2-proxy/oauth2-proxy)）
+  或者根本无法正常工作的第三方解决方案进行集成。
+
+- 授权策略缺少在您场景中所需的语义内容。
+
+## 解决方案 {#solution}
+
+在 Istio 1.9 中，我们通过引入
+[`CUSTOM` 操作](/zh/docs/reference/config/security/authorization-policy/#AuthorizationPolicy-Action)实现了授权策略的可扩展性，
+它允许您将访问控制决策委托给外部授权服务。
+
+`CUSTOM` 操作允许您将 Istio 与外部授权系统集成，
+该系统实现了自己的自定义授权逻辑。下图展示了此集成方式的顶层架构：
+
+{{< image width="100%"
+    link="./external_authz.svg"
+    caption="外部授权架构"
+    >}}
+
+在进行配置时，网格管理员使用 `CUSTOM` 操作对授权策略进行配置，
+用于在代理（网关或 Sidecar）上启用外部授权。
+管理员应确认外部身份验证服务已启动且正在运行。
+
+在运行时中：
+
+1. 请求被代理拦截，代理将根据用户在授权策略中的配置向外部授权服务发送检查请求。
+
+1. 外部授权服务将决定是否允许请求通过。
+
+1. 如果允许，请求将继续，并将由 `ALLOW`/`DENY` 操作定义的任意本地授权强制执行。
+
+1. 如果被拒绝，请求将立即被终止。
+
+让我们看一下带有 `CUSTOM` 操作的示例授权策略：
+
+{{< text yaml >}}
+apiVersion: security.istio.io/v1beta1
+kind: AuthorizationPolicy
+metadata:
+  name: ext-authz
+  namespace: istio-system
+spec:
+  # 选择器适用于 istio-system 命名空间中的入口网关。
+  selector:
+    matchLabels:
+      app: istio-ingressgateway
+  # “CUSTOM” 操作将访问控制委托给外部授权程序，
+  # 这与在代理内部强制执行访问控制的 ALLOW/DENY 操作不同。
+  action: CUSTOM
+  # provider 指定在 meshconfig 中定义的外部授权者的名称，
+  # 它告知我们在哪里以及如何与外部 auth 服务对话。稍后我们将对此进行更多介绍。
+  provider:
+    name: "my-ext-authz-service"
+  # 该规则指定只有当请求路径有前缀“/admin/”时才会触发访问控制。
+  # 这使您可以轻松地根据请求本身启用或禁用外部授权，避免在不需要时进行额外的外部检查请求。
+  rules:
+  - to:
+    - operation:
+        paths: ["/admin/*"]
+{{< /text >}}
+
+它引用了一个名为 `my-ext-authz-service` 的提供程序，
+该提供程序被定义在网格配置中：
+
+{{< text yaml >}}
+extensionProviders:
+# 名称“my-ext-authz-service”在其授权策略的 provider 字段中被引用。
+- name: "my-ext-authz-service"
+  # “envoyExtAuthzGrpc”字段指定外部授权服务的类型由 Envoy ext-authz filter gRPC API 实现。
+  # 另一种被支持的类型是 Envoy ext-authz 过滤器的 HTTP API。
+  # 在 https://www.envoyproxy.io/docs/envoy/v1.16.2/intro/arch_overview/security/ext_authz_filter 中查看更多信息。
+  envoyExtAuthzGrpc:
+    # service 和 port 字段指定外部认证服务的地址，
+    # “ext-authz.istio-system.svc.cluster.local”表示该服务部署在网格中。
+    # 它也可以是定义为在网格外部的服务，或者是在相同 Pod 内定义的独立容器。
```

Comment:
```suggestion
    # The service and port specifies the address of the external auth service, "ext-authz.istio-system.svc.cluster.local"
    # means the service is deployed in the mesh. It can also be defined out of the mesh or even inside the pod as a separate
    # container.
```
Same as above